go/kbfs/libkbfs/conflict_resolver.go

```diff
@@ -3183,6 +3194,121 @@ outer:
+	Version int
+	time.Time
+	Merged      string
+	Unmerged    string
+	ErrorTime   time.Time
+	ErrorString string
+	PanicString string
+	codec.UnknownFieldSetHandler
+}
+
+func (cr *ConflictResolver) recordStartResolve(ci conflictInput) error {
+	db := cr.config.KBFSOps().GetConflictResolutionDB()
+	conflictsSoFarSerialized, err := db.Get([]byte(cr.fbo.id().String()), nil)
+	var conflictsSoFar []conflictRecord
+	switch err {
+	case leveldb.ErrNotFound:
+		conflictsSoFar = nil
+	case nil:
+		err = cr.config.Codec().Decode(conflictsSoFarSerialized,
+			&conflictsSoFar)
+		if err != nil {
+			return err
+		}
+	default:
+		return err
+	}
+	if len(conflictsSoFar) > 10 {
+		conflictsSoFar = conflictsSoFar[len(conflictsSoFar)-10:]
+	}
+	conflictsSoFar = append(conflictsSoFar, conflictRecord{
+		Version:  conflictRecordVersion,
+		Time:     cr.config.Clock().Now(),
+		Merged:   ci.merged.String(),
+		Unmerged: ci.unmerged.String(),
+	})
+	conflictsSerialized, err := cr.config.Codec().Encode(conflictsSoFar)
+	if err != nil {
+		return err
+	}
+	return db.Put([]byte(cr.fbo.id().String()), conflictsSerialized, nil)
+}
+
+// recordFinishResolve does one of two things:
+//  - in the event of success, it deletes the DB entry that recorded conflict
+//    resolution attempts for this resolver
+//  - in the event of failure, it logs that CR failed and tries to record the
+//    failure to the DB.
+func (cr *ConflictResolver) recordFinishResolve(
+	ctx context.Context, ci conflictInput, receivedErr error) {
+	db := cr.config.KBFSOps().GetConflictResolutionDB()
+	panicVar := recover()
+
+	// If we neither errored nor panicked, this CR succeeded and we can wipe
+	// the DB entry.
+	if receivedErr == nil && panicVar == nil {
+		err := db.Delete([]byte(cr.fbo.id().String()), nil)
+		if err != nil {
+			cr.log.CWarningf(ctx,
+				"Could not record conflict resolution success: %v", err)
+		}
+		return
+	}
+
+	var err error
+	defer func() {
+		// If we can't record the failure to the CR DB, at least log it.
+		if err != nil {
+			cr.log.CWarningf(ctx,
+				"Could not record conflict resolution failure [%v/%v]: %v",
+				receivedErr, panicVar, err)
+		}
+		// If we recovered from a panic, keep panicking.
+		if panicVar != nil {
+			panic(panicVar)
+		}
+	}()
+
+	// Otherwise we need to decode the most recent entry, modify it, and put it
+	// back in the DB.
+	var conflictsSerialized []byte
+	conflictsSerialized, err = db.Get([]byte(cr.fbo.id().String()), nil)
+	var conflictsSoFar []conflictRecord
+	switch err {
+	case leveldb.ErrNotFound:
+		err = errors.New("No CR records")
+		return
+	case nil:
+		err = cr.config.Codec().Decode(conflictsSerialized,
+			&conflictsSoFar)
+		if err != nil {
+			return
+		}
+	default:
+		return
+	}
+
+	thisCR := &conflictsSoFar[len(conflictsSoFar)-1]
+	thisCR.ErrorTime = cr.config.Clock().Now()
+	if receivedErr != nil {
+		thisCR.ErrorString = receivedErr.Error()
+	}
+	if panicVar != nil {
+		thisCR.PanicString = fmt.Sprintf("%s", panicVar)
```

Comment:
It would also be great to get the full stack here.  According to https://groups.google.com/forum/#!topic/golang-nuts/MB8GyW5j2UY, `debug.Stack()` should give you the right thing, so maybe we can append it to the `panicVar` string?